File: child_android/app/src/main/kotlin/com/guardnest/kid/ScreenGuard.kt
package com.guardnest.kid

import android.content.Context
import java.util.Calendar

/**
 * Screen-time enforcement: pause and bedtime. Bedtime is evaluated by local
 * clock time so it works fully offline. When the device should be locked
 * (paused, or inside the bedtime window) we lock the screen.
 *
 * Times are "minutes from midnight" (0..1439), matching the parent app's rule
 * model, so the two sides speak the same language.
 */
object ScreenGuard {

    /** True while pause or bedtime is in effect (all apps blocked except
     *  emergency calls). Set by [EnforcementService], read by the accessibility
     *  service to bounce the child out of non-essential apps in real time. */
    @Volatile var locked = false

    /** "paused" / "resting (bedtime)" — for the on-screen message. */
    @Volatile var label = "paused"

    /** Title + message for the lock screen shown when the child opens an app
     *  during pause/bedtime. Set by [EnforcementService] so both the enforcement
     *  loop and the accessibility service show the same text (with the real
     *  bedtime schedule). */
    @Volatile var lockTitle = "Paused"
    @Volatile var lockSubtitle = "Your device is paused by your parent."

    /** True if [now] (minutes from midnight) is within [start, end), handling
     *  overnight windows like 21:00 -> 07:00. */
    fun isWithinWindow(now: Int, start: Int, end: Int): Boolean {
        if (start == end) return false
        return if (start < end) now in start until end else (now >= start || now < end)
    }

    fun nowMinutes(): Int {
        val c = Calendar.getInstance()
        return c.get(Calendar.HOUR_OF_DAY) * 60 + c.get(Calendar.MINUTE)
    }

    /**
     * Decides whether the device should currently be locked, from the family's
     * screen-time rule.
     */
    fun shouldLock(
        paused: Boolean,
        bedtimeEnabled: Boolean,
        bedtimeStart: Int,
        bedtimeEnd: Int,
        now: Int = nowMinutes(),
    ): Boolean {
        if (paused) return true
        if (!bedtimeEnabled) return false
        return isWithinWindow(now, bedtimeStart, bedtimeEnd)
    }

    /** Applies the lock decision: locks the screen now when it should be locked. */
    fun enforce(
        ctx: Context,
        paused: Boolean,
        bedtimeEnabled: Boolean,
        bedtimeStart: Int,
        bedtimeEnd: Int,
    ): String {
        val lock = shouldLock(paused, bedtimeEnabled, bedtimeStart, bedtimeEnd)
        return if (lock) DeviceLockdown.lockNow(ctx) else "Unlocked (rules allow use)."
    }
}

File: child_windows/src/GuardNest.Core/Rules/ScreenGuard.cs
namespace GuardNest.Core.Rules;

/// <summary>
/// Screen-time enforcement, ported from ScreenGuard.kt so both platforms lock at
/// exactly the same moments. Pure decisions, evaluated against the local clock so
/// they hold with no network.
/// </summary>
public static class ScreenGuard
{
    /// <summary>True when <paramref name="now"/> falls in [start, end), handling
    /// overnight windows like 21:00 -> 07:00.</summary>
    public static bool IsWithinWindow(int now, int start, int end)
    {
        if (start == end) return false;
        return start < end ? now >= start && now < end : now >= start || now < end;
    }

    public static int NowMinutes() => DateTime.Now.Hour * 60 + DateTime.Now.Minute;

    public static bool ShouldLock(ScreenTimeRule rule, int usedMinutesToday, int? now = null)
    {
        if (rule.Paused) return true;
        if (rule.DailyLimitMinutes > 0 && usedMinutesToday >= rule.DailyLimitMinutes) return true;
        if (!rule.BedtimeEnabled) return false;
        return IsWithinWindow(now ?? NowMinutes(), rule.BedtimeStart, rule.BedtimeEnd);
    }

    /// <summary>The two lines the lock screen shows, so every surface agrees.</summary>
    public static (string Title, string Subtitle) LockText(ScreenTimeRule rule, int usedMinutesToday)
    {
        if (rule.Paused)
        {
            return ("Paused", "Your device is paused by your parent.");
        }
        if (rule.DailyLimitMinutes > 0 && usedMinutesToday >= rule.DailyLimitMinutes)
        {
            return ("Screen time is up",
                $"You have used your {Describe(rule.DailyLimitMinutes)} for today.");
        }
        return ("Bedtime", $"Your device rests from {Clock(rule.BedtimeStart)} to {Clock(rule.BedtimeEnd)}.");
    }

    private static string Clock(int minutes) =>
        DateTime.Today.AddMinutes(minutes).ToString("h:mm tt");

    private static string Describe(int minutes) => minutes switch
    {
        < 60 => $"{minutes} minutes",
        60 => "1 hour",
        _ when minutes % 60 == 0 => $"{minutes / 60} hours",
        _ => $"{minutes / 60}h {minutes % 60}m",
    };
}
